From eb33f29a00d73214ae5adadd3e3b70e85c5ef411 Mon Sep 17 00:00:00 2001
From: Jins Varghese <dcw476@motorola.com>
Date: Mon, 15 Oct 2012 18:28:40 +0530
Subject: [PATCH 1/1] System Server Force Close.

Rootcause:  Wrong CommandNum isbeing set because of threaded
implementation of rtsol command. This is resulting in IllegalStateException
in framework.
Patch provided  by QC via SR 982201 .Passed stablity rack testing

Change-Id: I7dc3ca819d5a71121dc4f4c185eb4c2176e81f3e
Change-Id: I4b59c5d94e07843e94581868d3db0434f2afdd56
Reviewed-on: http://gerrit.pcs.mot.com/487780
Tested-by: Jira Key <JIRAKEY@motorola.com>
Reviewed-by: Binesh Balasingh <binesh@motorola.com>
Reviewed-by: Jeffrey Carlyle <jeff.carlyle@motorola.com>
(cherry picked from commit 5a6f12ad04bd51328eb3659b9a7bac671d4189ee)

Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 libsysutils/include/sysutils/SocketClient.h |  1 +
 libsysutils/src/SocketClient.cpp            | 12 ++++++++++++
 2 files changed, 13 insertions(+)

diff --git a/libsysutils/include/sysutils/SocketClient.h b/libsysutils/include/sysutils/SocketClient.h
index 1004f0611..3547d490a 100644
--- a/libsysutils/include/sysutils/SocketClient.h
+++ b/libsysutils/include/sysutils/SocketClient.h
@@ -48,6 +48,7 @@ public:
     int sendMsg(int code, const char *msg, bool addErrno);
     int sendMsg(int code, const char *msg, bool addErrno, bool useCmdNum);
     int sendMsg(const char *msg);
+    int sendMsgWithCmdNum(int code, const char *msg, int cmdNum);
 
     // Provides a mechanism to send a response code to the client.
     // Sends the code and a null character.
diff --git a/libsysutils/src/SocketClient.cpp b/libsysutils/src/SocketClient.cpp
index 971f90821..bd1d5eac6 100644
--- a/libsysutils/src/SocketClient.cpp
+++ b/libsysutils/src/SocketClient.cpp
@@ -121,6 +121,18 @@ int SocketClient::sendBinaryMsg(int code, const void *data, int len) {
     return result;
 }
 
+int SocketClient::sendMsgWithCmdNum(int code, const char *msg, int cmdNum) {
+    char *buf;
+    int ret = 0;
+    ret = asprintf(&buf, "%d %d %s", code, cmdNum, msg);
+    /* Send the zero-terminated message */
+    if (ret != -1) {
+        ret = sendMsg(buf);
+        free(buf);
+    }
+    return ret;
+}
+
 // Sends the code (c-string null-terminated).
 int SocketClient::sendCode(int code) {
     char buf[4];
-- 
2.11.0

